packages/application-generic/src/usecases/add-job/merge-or-create-digest.usecase.ts

```diff
@@ -0,0 +1,183 @@
+      job,
+      digestKey,
+      digestValue,
+      digestMeta
+    );
+
+    switch (digestAction) {
+      case DigestCreationResultEnum.MERGED:
+        return await this.processMergedDigest(job);
+      case DigestCreationResultEnum.SKIPPED:
+        return await this.processSkippedDigest(job);
+      case DigestCreationResultEnum.CREATED:
+        return await this.processCreatedDigest(digestMeta, job);
+      default:
+        throw new ApiException('Something went wrong with digest creation');
+    }
+  }
+
+  @Instrument()
+  private async processCreatedDigest(
+    digestMeta: IDigestBaseMetadata | undefined,
+    job: JobEntity
+  ): Promise<DigestCreationResultEnum> {
+    const regularDigestMeta = digestMeta as IDigestRegularMetadata | undefined;
+    if (!regularDigestMeta?.amount || !regularDigestMeta?.unit) {
+      throw new ApiException(
+        `Somehow ${job._id} had wrong digest settings and escaped validation`
+      );
+    }
+
+    return DigestCreationResultEnum.CREATED;
+  }
+
+  @Instrument()
+  private async processMergedDigest(
+    job: JobEntity
+  ): Promise<DigestCreationResultEnum> {
+    await this.jobRepository.update(
+      {
+        _environmentId: job._environmentId,
+        _id: job._id,
+      },
+      {
+        $set: {
+          status: JobStatusEnum.MERGED,
+        },
+      }
+    );
+
+    await this.jobRepository.updateAllChildJobStatus(job, JobStatusEnum.MERGED);
+
+    await this.digestMergedExecutionDetails(job);
+
+    return DigestCreationResultEnum.MERGED;
+  }
+
+  @Instrument()
+  private async processSkippedDigest(
+    job: JobEntity
+  ): Promise<DigestCreationResultEnum> {
+    await this.jobRepository.update(
+      {
+        _environmentId: job._environmentId,
+        _id: job._id,
+      },
+      {
+        $set: {
+          status: JobStatusEnum.SKIPPED,
+        },
+      }
+    );
+
+    return DigestCreationResultEnum.SKIPPED;
+  }
+
+  private getLockKey(
+    job: JobEntity,
+    digestKey: string,
+    digestValue: string | number
+  ): string {
+    let resource = `environment:${job._environmentId}:template:${job._templateId}:subscriber:${job._subscriberId}`;
+    if (digestKey && digestValue) {
+      resource = `${resource}:digestKey:${digestKey}:digestValue:${digestValue}`;
+    }
+
+    return resource;
+  }
+
+  private async shouldDelayDigestOrMergeWithLock(
+    job: JobEntity,
+    digestKey?: string,
+    digestValue?: string | number,
+    digestMeta?: any
+  ): Promise<DigestCreationResultEnum> {
+    const TTL = 1500;
+    const resourceKey = this.getLockKey(job, digestKey, digestValue);
+
+    const shouldDelayDigestJobOrMerge = async () =>
+      this.jobRepository.shouldDelayDigestJobOrMerge(
+        job,
+        digestKey,
+        digestValue,
+        digestMeta
+      );
+
+    const result =
+      await this.eventsDistributedLockService.applyLock<DigestCreationResultEnum>(
```

Comment:
Not sure if with the new logic flow we still need the Distributed Lock as we are handling the concurrency at DAL level by pessimistic locking. It would be great if we could finally get rid of this.

Reply:
Me neither but let's remove it in another PR. So we can keep this to fix the tests we discussed in the planning meeting.

Reply:
https://linear.app/novu/issue/NV-2870/investigate-if-distributed-lock-is-still-needed